xprof/sync_daemon_mpi.c

```diff
@@ -1,27 +1,94 @@
-#include <mpi.h>
+#include "mpi.h"
 #include <signal.h>
 #include <stdbool.h>
+#include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 
 // Define real-time signals
 #define RT_SIGNAL_READY SIGRTMIN
 #define RT_SIGNAL_GLOBAL_BARRIER SIGRTMIN + 1
 #define RT_SIGNAL_LOCAL_BARRIER SIGRTMIN + 2
 #define RT_SIGNAL_FINISH SIGRTMIN + 3
 
-int main(int argc, char **argv) {
+int MPIX_Init_Session(MPI_Session *lib_shandle, MPI_Comm *lib_comm) {
 
-  // Initialization
-  MPI_Init(&argc, &argv);
+  int rc, flag, valuelen;
+  int ret = MPI_SUCCESS;
+  const char pset_name[] = "mpi://WORLD";
+  const char mt_key[] = "thread_level";
+  const char mt_value[] = "MPI_THREAD_SINGLE";
+  char out_value[100]; /* large enough */
+  MPI_Group wgroup = MPI_GROUP_NULL;
+  MPI_Info sinfo = MPI_INFO_NULL;
+  MPI_Info tinfo = MPI_INFO_NULL;
+  MPI_Info_create(&sinfo);
+  MPI_Info_set(sinfo, mt_key, mt_value);
+  rc = MPI_Session_init(sinfo, MPI_ERRORS_RETURN, lib_shandle);
+  if (rc != MPI_SUCCESS) {
+    ret = -1;
+    goto fn_exit;
+  }
+  /*
+   * check we got thread support level foo library needs
+   */
+  rc = MPI_Session_get_info(*lib_shandle, &tinfo);
+  if (rc != MPI_SUCCESS) {
+    ret = -1;
+    goto fn_exit;
+  }
+  valuelen = sizeof(out_value);
+  MPI_Info_get_string(tinfo, mt_key, &valuelen, out_value, &flag);
+  if (flag == 0) {
+    fprintf(stderr, "THAPI_SYNC_DAEMON_MPI Warning: Could not find key %s\n", mt_key);
+    //ret = -1;
+    //goto fn_exit;
+  }
+  if (strcmp(out_value, mt_value)) {
+    fprintf(stderr, "THAPI_SYNC_DAEMON_MPI Warning: Did not get MPI_THREAD_SINGLE, got %s\n", out_value);
+    //ret = -1;
+    //goto fn_exit;
+  }
```

Comment:
Yep, but it's triggered in the CI :( So if we want the CI to pass, this is the best i can think off. 
We don't really care as `MPI_THREAD_SINGLE` is the mode with the most constrain, and the MPI cression session worked. So that mean we got a session just maybe more relatexed (aka can call MPI for multiple thread)